### PR TITLE
CONTROLLERS: POST Comment 

### LINKED ISSUE #75: CONTROLLERS: POST Comment

### PR DESCRIPTION
**Pull request by cjdutoit**
closes #75
image
image
image
image

### CODE REVIEW

File: Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            migrationBuilder.AddColumn<DateTimeOffset>(
+                name: "CreatedDate",
+                table: "Comments",
+                type: "datetimeoffset",
+                nullable: false,
+                defaultValue: new DateTimeOffset(new DateTime(1, 1, 1, 0, 0, 0, 0, DateTimeKind.Unspecified), new TimeSpan(0, 0, 0, 0, 0)));
+
+            migrationBuilder.AddColumn<DateTimeOffset>(
+                name: "UpdatedDate",
+                table: "Comments",
```
Comment (hassanhabib) [inline Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs:21]: These changes don't seem to be related to a controller task

File: Taarafo.Core/Controllers/CommentsController.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    [Route("api/[controller]")]
+    public class CommentsController : RESTFulController
+    {
+        private readonly ICommentService commentService;
+
+        public CommentsController(ICommentService commentService) =>
+            this.commentService = commentService;
+
+        [HttpPost]
+        public async ValueTask<ActionResult<Comment>> CommentCommentAsync(Comment comment)
```
Comment (ShriHumrudha) [inline Taarafo.Core/Controllers/CommentsController.cs:25]: PostCommentAsync than CommentCommentAsync

File: Taarafo.Core/Controllers/CommentsController.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            }
+            catch (CommentValidationException commentValidationException)
+            {
+                return BadRequest(commentValidationException.InnerException);
+            }
+            catch (CommentDependencyValidationException commentDependencyValidationException)
+               when (commentDependencyValidationException.InnerException is AlreadyExistsCommentException)
+            {
+                return Conflict(commentDependencyValidationException.InnerException);
+            }
```
Comment (ShriHumrudha) [inline Taarafo.Core/Controllers/CommentsController.cs:42]: We need another exception - InvalidCommentReferenceException

**Comment by ShriHumrudha** [comment]
> closes #75 image

Do share the screenshots for Bad Request - 400, Conflict-409, Failed Dependency - 404

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs (hunk 1 above)
Comment (cjdutoit) [reply Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs:21]: @hassanhabib I had to add this to get the controller to return a valid output as the migrations was not up to date.

File: Taarafo.Core/Controllers/CommentsController.cs (hunk 3 above)
Comment (cjdutoit) [reply Taarafo.Core/Controllers/CommentsController.cs:50]: @ShriHumrudha I have added this and updated the screenshots to include Bad Request - 400, Conflict-409, Failed Dependency - 424

File: Taarafo.Core/Controllers/CommentsController.cs (hunk 2 above)
Comment (cjdutoit) [reply Taarafo.Core/Controllers/CommentsController.cs:26]: @ShriHumrudha I corrected this

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs [hunk 4]
```diff
@@ -369,5 +370,67 @@ public async Task ShouldThrowValidationExceptionOnModifyIfStorageUpdatedDateSame
...
+            // given
+            int minuteInPast = GetRandomNegativeNumber();
+            DateTimeOffset randomDate = GetRandomDateTimeOffset();
+            Comment randomComment = CreateRandomModifyComment(randomDate.AddMinutes(minuteInPast));
+            Comment inputComment = randomComment.DeepClone();
+            inputComment.UpdatedDate = randomDate;
+            Comment storageComment = randomComment;
+            string randomMessage = GetRandomMessage();
+            string exceptionMessage = randomMessage;
+            var foreignKeyConstraintConflictException = new ForeignKeyConstraintConflictException(exceptionMessage);
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs:386]: New line space.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs [hunk 5]
```diff
@@ -369,5 +370,67 @@ public async Task ShouldThrowValidationExceptionOnModifyIfStorageUpdatedDateSame
...
+
+            var invalidCommentReferenceException =
+                new InvalidCommentReferenceException(foreignKeyConstraintConflictException);
+
+            var expectedCommentValidationException =
+                new CommentValidationException(invalidCommentReferenceException);
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs:392]: This will be dependency validation exception and goes under 'Exceptions' file

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs [hunk 6]
```diff
@@ -369,5 +370,67 @@ public async Task ShouldThrowValidationExceptionOnModifyIfStorageUpdatedDateSame
...
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                    .Returns(randomDate);
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs:396]: We shall throw the exception in the first dependency call.

File: Taarafo.Core/Controllers/CommentsController.cs (hunk 2 above)
Comment (ShriHumrudha) [reply Taarafo.Core/Controllers/CommentsController.cs:26]: I still don't see the change here.

File: Taarafo.Core/Controllers/CommentsController.cs (hunk 3 above)
Comment (ShriHumrudha) [reply Taarafo.Core/Controllers/CommentsController.cs:50]: Please verify if you have pushed the code.

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs (hunk 1 above)
Comment (hassanhabib) [reply Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs:21]: @cjdutoit let's have a different pr for this then. `DATA: Add Created & Updated Dates`

Comment (cjdutoit) [reply Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs:21]: Submitted PR #78 to address this.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs (hunk 6 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs:396]: @ShriHumrudha I will update,  can you explain why we throw this there?   I will also have to update this on the other one as that is also done on the storageBroker rather than the 1st one.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs (hunk 5 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs]: @ShriHumrudha, I have updated this

File: Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs (hunk 1 above)
Comment (hassanhabib) [reply Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs:21]: @cjdutoit perfect. thanks. I just left you one tiny comment there.

**Comment by hassanhabib** [comment]
Holding off the merge of this PR until https://github.com/hassanhabib/Taarafo.Core/pull/78 is merged in.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs [hunk 7]
```diff
@@ -369,5 +370,67 @@ public async Task ShouldThrowValidationExceptionOnModifyIfStorageUpdatedDateSame
             this.dateTimeBrokerMock.VerifyNoOtherCalls();
             this.loggingBrokerMock.VerifyNoOtherCalls();
         }
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnModifyWhenReferneceExceptionAndLogItAsync()
```
[the remark this answers is not in the document]
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs]: Renamed

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs (hunk 4 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validation.Modify.cs]: Fixed

File: Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs (hunk 1 above)
Comment (cjdutoit) [reply Taarafo.Core/Migrations/20211110140549_UpdateCommentModel.cs]: closed via #78

**Comment by ShriHumrudha** [comment]
@cjdutoit Thank you so much for the effort. I see there are multiple things clubbed in one PR. The scope of this one is just CONTROLLERS, but I see foundation & service registration are clubbed too.

@hassanhabib - Please share your thoughts.

**Comment by hassanhabib** [comment]
> @cjdutoit Thank you so much for the effort. I see there are multiple things clubbed in one PR. The scope of this one is just CONTROLLERS, but I see foundation & service registration are clubbed too.
> 
> @hassanhabib - Please share your thoughts.

@ShriHumrudha I agree. @cjdutoit break everything that is not a controller in this PR into it's own PR please.

**Comment by cjdutoit** [comment]
@hassanhabib @ShriHumrudha I will split them up into different PS's  I thought these would be OK on this PR since those exceptions / code were added as a result of the review on this code.  Also, we might need to review gaps in the Posts foundation service and controller again as I used that as a guide for this work.

**Comment by hassanhabib** [comment]
@cjdutoit sounds good thanks.
